### ISSUE TITLE
Historical view

### ISSUE THREAD

**Issue by andreva7** [clovett/MyMoney.Net#28]
Add support to view the value of the whole portfolio and accounts at a date in the past.

A dynamic implementation would likely be very hard. But a simple solution that solves the immediate problem would be to support a filter date when opening a database. When a date is set, all transactions after that date are simply not loaded. The other feature necessary would be to set all the stock prices based o that filter date. That would work as I believe we keep the whole stock price history for each stock.
As a safety, this mode could possibly automatically disable the save option so people don't accidentally through away years of recent transactions.

**Comment by clovett** [comment]
Great idea, I've wanted to do this also, should not be difficult now that we have stock quote history and we already have a class that can compute the value of the portfolio at a given date based on this, just need the UI for it...

**Comment by andreva7** [comment]
Would still be very useful to me as I am working on financial deliverables to my lawyer

**Comment by clovett** [comment]
See https://github.com/clovett/MyMoney.Net/pull/53

Let me know if you can try it out before I publish a new version in case there's any bugs...

@jpdup might be interested in this PR also.

**Comment by andreva7** [comment]
#28 was to solve a different problem.

For example, I want to see how many shares I owned in an account as a particular date.
The bigger picture, which I what I need, is to show all the shares I owned and all the cash I had across all accounts on a particular date. Essentially, net worth, on a specific day in the past. But then I want to see it by account also. And ideally, I don't want see anything that happened after that date.

Basically, I want to see the state of the world on some day in 2012.

I think the simple approach is to trim all transactions that happen after that day.
A cooler approach would be to create a filter on the transaction reader, but that may be complicated to plum across all the code that enumerates through all transactions, unless there is a way to put that deep in the enumeration method.

-Andre

**Comment by lovettchris** [comment]
Yes, I don't have it across all accounts, but the new TrendGraph on investment accounts, if you export it to .csv will give you the exact market value of the account on each date you care to look at in that .csv file.

**Comment by clovett** [comment]
But yes, I can add some options on the Reports so that you can view the state of your portfolio at a given date...

**Comment by clovett** [comment]
@andreva7, See https://github.com/clovett/MyMoney.Net/pull/54 for a work in progress... if you have a way of coroborating the numbers reported there with some other documents you have it would be a great test.

**Comment by andreva7** [comment]
Debugging without looking at the code … 😊

I am guessing the networth report start with all accounts with a non-zero value, and then figures out their value at that earlier date.
I have many accounts that have zero value today, but had a value back then.

-Andre

**Comment by clovett** [comment]
It always has to start with the oldest transactions and move forwards so that it can apply stock splits at the right time on the remaining units at the point of the stock split, the change in the PR then is to simply stop at the given date rather than going all the way through to today, then also lookup the stock prices for that date rather than using the latest stock prices.

**Comment by andreva7** [comment]
I bought a house in 2010, sold it in 2021. The value of that account is now $0.
When I select net worth for 2012, the house does not show up on the list of Assets.

**Comment by clovett** [comment]
I just pushed a possible fix, there was a check that the asset account balance was > 0, it now checks >= 0

**Comment by jpdup** [comment]
Yes nice feature to have

**Comment by clovett** [comment]
@jpdup I pushed some fixes to how I was computing historical value of loan accounts, I had to refactor some stuff in Money_loan.cs.  I added a new abstraction named "Loan" so that I can avoid having to do the very expensive GetLoanPaymentsAggregation too many times.

**Comment by jpdup** [comment]
@clovett thanks for the refactor

**Comment by andreva7** [comment]
Looks good

**Comment by andreva7** [comment]
I am testing with this change.

I am comparing this version vs the published version. For some reason, without applying any filter, the two versions report a different number for "liabilities ->credit". I am trying to go through all the balances of my accounts to figure out if there is another liability somewhere, which is where I hit the "balance column" missing.

-Andre
